### ISSUE TITLE
memset/memcpy for mapped buffers in WASM

### ISSUE THREAD

**Issue by kainino0x** [webgpu-native/webgpu-headers#66]
@juj on the implementation of these two functions, which involve a `malloc`, `Uint8Array.set`, and `free`.

> Oh, these two functions `wgpuBufferGetConstMappedRange` and `wgpuBufferGetMappedRange` are horrible 🤮 😞
> 
> I suppose mapping a GPU buffer directly to wasm heap address space will be a hard no-go? (has that been discussed with Wasm folks? CC @lukewagner @dschuff )
> 
> Assuming it is not possible, I wonder if we should consider changing up the API completely. Given that there is only asynchronous mapping support, we should make this a callback-based API where a cb function gets called when the map happens, and the cb function will receive a handle ID, and then have dedicated functions
> 
> ```js
> void wgpuMemset8ToMappedBuffer(uint destinationMappedBufferHandleId, uint destinationOffset, uint destinationSize, uchar fillValue);
> void wgpuMemset16ToMappedBuffer(uint destinationMappedBufferHandleId, uint destinationOffset, uint destinationSize, ushort fillValue);
> void wgpuMemset32ToMappedBuffer(uint destinationMappedBufferHandleId, uint destinationOffset, uint destinationSize, uint fillValue);
> void wgpuMemcpyToMappedBuffer(uint destinationMappedBufferHandleId, uint destinationOffset, uint destinationSize, void *sourcePtr);
> void wgpuMemcpyFromMappedBuffer(void *destinationPtr, uint destinationOffset, uint destinationSize, uint sourceMappedBufferHandleId);
> ```
> 
> This way the caller will be in control whether there will need to be malloc/free in effect, and avoid any redundant memsets and memcpys?

**Comment by austinEng** [comment]
I think we could add these type of idea with a webgpu-native extension and not need to change the JS API.

Instead of `wgpuBufferGetMappedRange`, we have `wgpuBufferGetMappedRangeHandle`, and add his suggestions. This keeps the mapped range js-side without needing to copy it into the Emscripten heap. Then we can add his helper suggestions which work with the handle.

Having it be callback-based doesn't seem relevant to the extra memory operations. I'm guessing that part is because he's concerned about garbage?

**Comment by kainino0x** [comment]
We already have mapAsync() for asynchronicity and unmap() for memory management; I don't think any more callbacks are needed.

I was thinking we would just put these helpers next to wgpuBufferGetMappedRange and each one would do a getMappedRange, but I guess since you might need to memcpy multiple times to one region (e.g. one chunk at a time) it would be better to have a handle instead.

That said, I really just want wasm mmapping.

**Comment by Kangz** [comment]
I think the intent was that creating a mapping a buffer or creating a buffer mapped at creation in WASM would do a malloc on map, and  writeBuffer + free on unmap(), at least for the `GPUMapMode.WRITE` case.

WASM mmaping would be best though.

**Comment by kainino0x** [comment]
Hm, that's an interesting idea. If the user is just memcpying into the mapped mem, then compared with the user using writeBuffer directly it's still one extra copy, but compared with the current implementation I think it's 0-1 fewer, depending on the browser implementation?

**Comment by Kangz** [comment]
Yeah mapping is only as good as `writeBuffer` if the application can "unpack" data in the mapped range. I think it's probably ok because the best we can do in WASM with the current API and WASM capabilities is `writeBuffer`.

**Comment by kainino0x** [comment]
We (well, @lokokung) have been working on updating the Emscripten bindings implementation to the stable C header and of course we've come to this topic again. I think we will explore adding methods like the ones described above (at least the memcpy ones) as Emscripten extensions, but we might want to just make them core for portability (even though they're not really useful in native).

Underneath these would call `getMappedRange`, so on Wasm they couldn't be overlapped with any other calls to these methods or `wgpuBufferGet(Const)MappedRange`. There are ways around this and we may be able to lift the restriction in the future (see discussion on #485) but for now this is the simplest to implement. The names I have in mind are:
`WGPUStatus wgpuBuffer[Emscripten]WriteMappedRange(WGPUBuffer, size_t offset, void const * data, size_t size)`
`WGPUStatus wgpuBuffer[Emscripten]ReadMappedRange(WGPUBuffer, size_t offset, void * data, size_t size)`

FWIW I don't think it is practical to replace the core `Get(Const)MappedRange` APIs because sometimes you really need a pointer, and that is efficient on native and probably eventually will be on wasm too.

**Comment by kainino0x** [comment]
Jan 9 meeting:

* WASM can’t directly access the buffer mappings because they aren’t in the WASM memory. Basically requires a memcpy to make things work at the moment.  
* KN: We will probably write these as Emscripten extensions anyways. Eventually maybe we won’t need them, but will be a while.  
* KN: We could put it in core? But we can also just start with it as an extension.  
* KN: The exact rules need to be figured out because JS doesn’t allow overlapped ranges.  
* CF: May as well have them in core  
* **Put them in core**   
* LK: Right now if we GetMappedRange on the same range in native, it's valid  
* KN: Header spec disallows writable overlaps (Dawn needs to update). Wasm disallows all overlaps. Originally disallowed writable overlaps because safe Rust API wouldn't allow it.  
  * LK: Might break applications, need to try and see  
  * CF: It should be possible for wgpu-native to return overlapping mutable ranges too, it's unsafe in the end anyway  
* Possible to workaround by allocating the whole mapAsync region up front, and give pointers into it, but inefficient.  
* (Discussion of JS allowing multiple mapAsync at the same time; it's wrapped up in feature requests for using one range mapped and one range on the GPU. And all that is kinda blocked on mapping for UMA.)  
* Anyway, about overlap validation:  
* LK: It is possible to make WriteMappedRange/ReadMappedRange work without overlap validation, by breaking up the memcpy into chunks  
* LK: If we allocate the whole mapAsync region up front, we would have to have WriteMappedRange use that, which defeats the purpose.  
* KN: It would be much simpler if we just had the same overlap validation on native  
* **all: OK to disallow all overlaps in native for now. We can always loosen it later.**  
* **LK: We'll have to prototype in Dawn to make sure we're not breaking anyone.**  
  * (same as with overlap checking for writable ranges anyway)  
  * Maybe we'll discover a use case that motivates changing it back.

**Comment by kainino0x** [comment]
In case it's not clear from the notes, we only talked about the memcpy versions (WriteMappedRange and ReadMappedRange). The memset versions are of interest too but we're keeping it to the known-critical ones for now.

**Comment by kainino0x** [comment]
Opened #494. Will leave this open for memset or other future improvements.
